server.py

```diff
@@ -1,69 +1,132 @@
-User = namedtuple("User", "password")
+db = SqliteDatabase(DATABASE_NAME)
+
+
+class User(Model):
+    username = CharField(unique=True)
+    password = CharField()
+
+    class Meta:
+        database = db
+
+
+class Session(Model):
+    id = CharField(unique=True)
+    user = ForeignKeyField(User)
+
+    class Meta:
+        database = db
+
+
+class Contest(Model):
+    code = CharField(unique=True)
+    description = CharField()
+    start_time = DateTimeField()
+    end_time = DateTimeField()
+
+    class Meta:
+        database = db
+
+
+class Question(Model):
+    q_no = IntegerField(unique=True)
+    author = ForeignKeyField(User)
+
+    class Meta:
+        database = db
+
+
+class ContestProblems(Model):
+    contest = ForeignKeyField(Contest)
+    question = ForeignKeyField(Question)
+
+    class Meta:
+        database = db
+        indexes = ((("contest", "question"), True),)
+
+
+class Submission(Model):
+    user = ForeignKeyField(User)
+    time = DateTimeField()
+    question = ForeignKeyField(Question)
+    contest = ForeignKeyField(Contest)
+    is_correct = BooleanField()
+
+    class Meta:
+        database = db
+        indexes = ((("user", "time"), True),)
+
+
+db.connect()
+db.create_tables([User, Session, Submission, ContestProblems, Contest, Question])
 
-# dummy contests
-contests["PRACTICE"] = Contest(
+
+# dummy contest data
+practiceContest = Contest.get_or_create(
+    code="PRACTICE",
     description="practice questions",
-    questions=[1, 2],
     start_time=datetime.datetime(day=1, month=1, year=1),
     end_time=datetime.datetime(day=1, month=1, year=9999),
 )
-contests["PASTCONTEST"] = Contest(
+pastContest = Contest.get_or_create(
+    code="PASTCONTEST",
     description="somewhere in the past",
-    questions=[1, 2],
     start_time=datetime.datetime(day=1, month=11, year=2018),
     end_time=datetime.datetime(day=1, month=12, year=2018),
 )
-contests["ONGOINGCONTEST"] = Contest(
+ongoingContest = Contest.get_or_create(
+    code="ONGOINGCONTEST",
     description="somewhere in the present",
-    questions=[3, 4],
     start_time=datetime.datetime(day=1, month=4, year=2019),
     end_time=datetime.datetime(day=1, month=6, year=2019),
 )
-contests["FUTURECONTEST"] = Contest(
+futureContest = Contest.get_or_create(
+    code="FUTURECONTEST",
     description="somewhere in the future",
-    questions=[5, 6],
     start_time=datetime.datetime(day=1, month=1, year=2020),
     end_time=datetime.datetime(day=1, month=10, year=2020),
 )
 
-for i in os.listdir(question_dir):
-    if not i.isdigit():
-        continue
-    with open(os.path.join(question_dir, i, "output.txt"), "rb") as fl:
-        output = fl.read()
-    with open(os.path.join(question_dir, i, "statement.txt"), "r") as fl:
-        statement = fl.read()
-    questions[i] = Question(output=output, statement=statement)
+test = User.get_or_create(username="test", password="test")
+
+q1 = Question.get_or_create(q_no=1, author=test[0])
```

Comment:
let's not use magic numbers. `test.username`

Reply:
On running `print(test)`, I get the output, `(<User: 1>, False)` (a tuple is returned)
While writing the code, I also tried variants like `test.username`, `test.User`, `test[username]` but none worked, then I finally managed with test[0], rest did not work for me.
I tried reading about tuples, they are accessed by indices only.

Reply:
When we add a system or making contests and questions interface, this can be dealt with as we would not have to add `get_or_create()` function

Reply:
:man_facepalming: right. I forgot that. Usually what we do at work is something like `test, _ = x.get_or_create()` which makes it clear that I don't care about the second thing, just the first. This is fine for now.